File: Assets/Scripts/zombieAi.js
#pragma strict
	public var attackInterval:float;
	private var timeSinceAttack:float = 0;
	public var attackSize:float;

	public var fearInterval:float;
	private var timeSinceFear:float;
	public var fearSize:float;

	//public var infectChance:float;
	
	public var damage:float;
	public var navAgent:NavMeshAgent ;
	private var currentTarget:GameObject;
	
	//private var currentAnimation;

	private var animComponent : Animation;
	
	public var lifeSpan:float;
	private var timeAlive : float =  0;
	
function Start () {
	//load NavAgent
		
		//Destroy(gameObject, lifeSpan);
		
		navAgent = GetComponent(NavMeshAgent);
		animComponent = GetComponent(Animation);
		
		//Pick a random target from the designated objects
		var g : GameObject = getRandomNavTarget ("SafeZone");		
		if (g == null) {g = getRandomNavTarget("Finish");}
		
		if (g != null) {
				setTarget(g);
		}
		//Go to new target
		//navAgent.SetDestination(currentTarget.transform.position);

		castFear();
		
		animComponent.wrapMode = WrapMode.Loop;
		animComponent.Play("walk");
		
}

//function PlayAnimation(animationName) {			
//			a.wrapMode = WrapMode.Loop;
//		
//			a.Play(animationName.ToString());
//			currentAnimation = animationName;
//							
//}


function FixedUpdate () {
//	if (navAgent.velocity.magnitude >0 && currentAnimation != "walk") {
//			PlayAnimation("LM_walk");
//			
//		}
	timeAlive += Time.deltaTime;
	
	timeSinceAttack += Time.deltaTime;
	timeSinceFear += Time.deltaTime;

	if (timeAlive >= lifeSpan) {
	die();
	}
		//check if time to attack again
	if (timeSinceAttack > attackInterval) {
		
		attack();
		timeSinceAttack = 0;

		//castFeat to scare other nearby humans
		//castFear();
	} else

	if (timeSinceFear > fearInterval) {
		castFear ();
		timeSinceFear = 0;
	}

	
						
		//If close to nav destination, send damage and die
//		if (navAgent.remainingDistance < 5 && currentTarget.tag == "SafeZone") {
		//	currentTarget.SendMessage("takeDamage", attackDamage);
			
		//	Destroy(gameObject);
			//currentTarget = getRandomNavTarget ("SafeZone");
			//navAgent.SetDestination (currentTarget.transform.position);
//		}


}




function getRandomNavTarget (tagName:String) {
		var c : gameControl   = Camera.main.GetComponent(gameControl);
		if (c.currentZombieTarget != null) {
			return c.currentZombieTarget;
		}	else {	

			var targets: GameObject[]  = GameObject.FindGameObjectsWithTag (tagName);

			if (targets.Length > 0 ) {
				var x:int = Random.Range (0, targets.Length);
		
				return targets [x];		
			}
		
		}
		
		return null;
	}
	
function setTarget (g:GameObject) {
	currentTarget = g;
	navAgent.SetDestination(currentTarget.transform.position);

	}	
	
//Destroy zombie object
function die() {
	//PlayAnimation("LM_die");
	
	navAgent.Stop();
	
	animComponent.wrapMode = WrapMode.Once;
	
	animComponent.Play("die");
	
	yield WaitForSeconds (animation["die"].length);
		
	Destroy (gameObject);
}	
	
//function attack() {
//						
//		var r:RaycastHit;
//
//		//sphere cast in front of zombie
//		if (Physics.SphereCast (new Ray (transform.position, transform.forward), attackSize,  r, 3)) {		
//				
//				//check if hits are human
//				if (r.collider.gameObject.tag == "Human") {		
//					//Destroy human
//					r.collider.gameObject.SendMessage("die");
//					//if (Random.Range(0f, 1f) <= infectChance) {
//					Camera.main.SendMessage("createZombie", r.collider.transform.position);
//				//}
//					
//				} else if (r.collider.gameObject.tag == "Guard") {		
//				//	Destroy(gameObject);
//				}
//				else if (r.collider.gameObject.tag == "SafeZone") {
//					r.collider.gameObject.SendMessage("takeDamage", attackDamage);
//					
//				}
//				
//
//		}
//	}

function attack() {
		//Debug.Log("Guard attacking!");
		
		
				
				
		var colliders : Collider[] = Physics.OverlapSphere(gameObject.transform.position, attackSize);
										
		for (var hit : Collider in colliders) {
			Debug.Log(hit.collider.gameObject);
			
			if (hit.gameObject.tag == "Human") {
				//Debug.Log("Zombie hit Human!");
				//hit.collider.gameObject.SendMessage("die");
				//engageTarget(hit.gameObject);
				//hit.collider.gameObject.SendMessage("takeDamage", damage);
				hit.gameObject.SendMessage("die");
					//if (Random.Range(0f, 1f) <= infectChance) {
					
				
				Camera.main.SendMessage("createZombie", hit.transform.position);
				
				animComponent.wrapMode = WrapMode.Once;
				animComponent.Play("L2R_swipe");
						
				animComponent.PlayQueued("walk");
				
				
				return;
			} else if (hit.gameObject.tag == "SafeZone") {
					hit.gameObject.SendMessage("takeDamage", damage);
			} else if (hit.gameObject.tag == "Guard") {
					hit.gameObject.SendMessage("takeDamage", damage);
					}		
		}
		
}

function stop() {
	navAgent.Stop();
}

	//send spherecast to find humans to scare
function castFear() {

		//get all hits in range
		for (var r:RaycastHit  in  Physics.SphereCastAll (new Ray (transform.position, transform.forward), fearSize, fearSize)) {

			//check if hits are uhman
			if (r.collider.gameObject.tag == "Human") {

				//tell human to run
				r.collider.gameObject.SendMessage("Afraid");
			}
				
		}
	}

function goingtoSafe() {
	if (currentTarget.tag == "SafeZone") 
		{return true;}
	else 
		{return false;}
}

File: Assets/Scripts/zombieAi.cs
using UnityEngine;
using System.Collections;
using System.Collections.Generic;

public class zombieAI : commonAI
{
	// default life-span time for a zombie.
	// self-die after given amount of time.
	public float lifeSpan;
	private float timeAlive = 0.0f;

	// Zombies can attack...
	// these scripts will also be added to the zombie at interface
	// time and we can adjust attack and cast fear properties there.
	private attackAI Attack;
	private castFearAI CastFear;
	
	public float healthDrainRate = 20.0f;

	public override void Start()
	{
		lifeSpan = 5000f;
		// do baseline start actions first...
		// also grabs default animation component too.
		base.Start();

		// upon first creation, create the common "attackAI" instance
		// Attack = this.gameObject.AddComponent<attackAI>();
		Attack = this.gameObject.GetComponent<attackAI>();

		// Now, for scaring humans (or whatever else may be scared away.
		// CastFear = this.gameObject.AddComponent<castFearAI>();
		CastFear = this.gameObject.GetComponent<castFearAI>();

		// What can a Zombie attack... Guards and Humans
		// and don't forget about the SafeZone houses too!
		Attack.AddAttackTarget(eNavTargets.Human);
		Attack.AddAttackTarget(eNavTargets.Guard);
		Attack.AddAttackTarget(eNavTargets.SafeZone);
		
		// When an attack is successful, we need to do something
		// very specific if attacking a human (change them into a zombie)
		// attach to the public exposed event handler created
		Attack.OnWasAttacked += HandleSpecificHitTarget;

		// via the "Hero Zombie" prefab, it has a child element
		// of "LittleHero_solo" which has animations of
		// L2R_swipe, walk, die, idle_lookaround, hurt
		Attack.AssignAnimation( gameObject, animComponent, "L2R_swipe" );


		// Zombies have same default targets of SafeZone and Finish
		// only directly target zombies to safe or finish zones.
		// however, why dont the zombies try to target humans and guards too
		// maybe later
		defaultNavTargets.Clear();
		defaultNavTargets.Add(eNavTargets.SafeZone);
		defaultNavTargets.Add(eNavTargets.Finish);

		// pick a new target from either safe vs finish possibilities
		moveToNewTarget();

		// Initiate the zombie to walking animation
		animComponent.wrapMode = WrapMode.Loop;
		animComponent.Play("walk");
	}
	
	public void setTarget( GameObject newTarget)
	{
		// In case object is being destroyed, don't allow set
		if( isDestroying )
			return;

		// in case issue with human-zombie conversion and navAgent lost
		if( navAgent != null )
			navAgent.SetDestination(newTarget.transform.position); 
	}

	// What is difference between Update and FixedUpdate???
	void FixedUpdate() 
	{
		// if time to die, get out, nothing more to do
		// zombies have a self-destruct time interval...
		if (timeToDie())
			return;

		//check if time to attack again
		Attack.CheckAttack();

		// check if zombie cast fear in human object
		CastFear.CheckScare();

		// check if target was reached, if so, get new target
		if( ! reachedTarget())
			// in addition, check for being stagnant (clarification in commonAI.cs)
			IsMovementStagnant();
	}

	// see if the zombie is time to die and self-destruct
	private bool timeToDie()
	{
		// if already marked as being destroyed, don't do it again...
		if (isDestroying)
			return true;

		timeAlive += Time.deltaTime;
		if (timeAlive >= lifeSpan)
			// NOW, destroy the zombie
			die();

		return isDestroying;
	}

	// When anything is attacked by a zombie, this method will be called.
	private void HandleSpecificHitTarget(Collider hit)
	{
		// ONLY if the object is a human do we create a newly spawned zombie.
		if (hit.gameObject.tag.Equals("Human"))
			Camera.main.SendMessage("createZombie", hit.transform.position);
	}
}
